File: tooling/gtk-bench/c/bench.c
// The floor: the same cases in C, calling GTK directly with no binding
// between. `run.sh` compiles it with -O2 and reports it beside nts and GJS,
// so each row says how much of its time is GTK's and how much is ours.
#include <gtk/gtk.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static double now(void) { return (double)g_get_monotonic_time() / 1e3; }

static void best(const char *name, long n, void (*run)(long)) {
  run(n);
  double fastest = 1e300;
  for (int rep = 0; rep < 3; rep++) {
    double start = now();
    run(n);
    double took = now() - start;
    if (took < fastest)
      fastest = took;
  }
  printf("%s %.1f\n", name, fastest * 1e6 / (double)n);
}

static long changes;
static double value;
static GtkAdjustment *adjustment;
static void changed(GtkAdjustment *a, gpointer data) {
  (void)a;
  (void)data;
  changes++;
}
static void signal_run(long n) {
  for (long i = 0; i < n; i++)
    gtk_adjustment_set_value(adjustment, ++value);
}

static GtkLabel *label;
static long length;
static void property_run(long n) {
  for (long i = 0; i < n; i++) {
    gtk_label_set_label(label, (i & 1) == 0 ? "a" : "bb");
    length += (long)strlen(gtk_label_get_label(label));
  }
}

static void construct_run(long n) {
  for (long i = 0; i < n; i++) {
    GtkWidget *made = g_object_ref_sink(g_object_new(GTK_TYPE_LABEL, NULL));
    gtk_label_set_label(GTK_LABEL(made), "x");
    if (strlen(gtk_label_get_label(GTK_LABEL(made))) != 1)
      puts("construct: wrong label");
    g_object_unref(made);
  }
}

/* The widget and the count are locals, as a C programmer writes the loop:
 * a static global is re-read after every call, since the call could have
 * written it, and that load and the counter's store were what put this
 * floor above nts's own loop, which holds the receiver in a register. */
static GtkWidget *button;
static long visible;
static void method_run(long n) {
  GtkWidget *widget = button;
  long seen = 0;
  for (long i = 0; i < n; i++)
    if (gtk_widget_get_visible(widget))
      seen++;
  visible += seen;
}

/* Out parameters: two ints written through slots on the stack. */
static long outs_sum;
static void outs_run(long n) {
  GtkWidget *widget = button;
  long sum = 0;
  for (long i = 0; i < n; i++) {
    int width, height;
    gtk_widget_get_size_request(widget, &width, &height);
    sum += width + height;
  }
  outs_sum += sum;
}

/* A widget class of the benchmark's own whose `measure` answers constants. */
typedef struct {
  GtkWidget parent;
} BenchSquare;
typedef struct {
  GtkWidgetClass parent_class;
} BenchSquareClass;
G_DEFINE_TYPE(BenchSquare, bench_square, GTK_TYPE_WIDGET)
static void bench_square_measure(GtkWidget *widget, GtkOrientation orientation,
                                 int for_size, int *minimum, int *natural,
                                 int *minimum_baseline, int *natural_baseline) {
  (void)widget;
  (void)orientation;
  (void)for_size;
  *minimum = 42;
  *natural = 42;
  *minimum_baseline = -1;
  *natural_baseline = -1;
}
static void bench_square_class_init(BenchSquareClass *klass) {
  GTK_WIDGET_CLASS(klass)->measure = bench_square_measure;
}
static void bench_square_init(BenchSquare *square) { (void)square; }

static GtkWidget *square;
static long measured;
static void vfunc_run(long n) {
  GtkWidget *widget = square;
  long total = 0;
  for (long i = 0; i < n; i++) {
    int size;
    gtk_widget_measure(widget, GTK_ORIENTATION_HORIZONTAL, 100 + (int)(i % 1000),
                       &size, NULL, NULL, NULL);
    total += size;
  }
  measured += total;
}

static void mapped(GtkWidget *w, gpointer app) {
  (void)w;
  g_application_quit(G_APPLICATION(app));
}
static void activate(GtkApplication *app, gpointer data) {
  (void)data;
  GtkWidget *window = gtk_window_new();
  gtk_window_set_title(GTK_WINDOW(window), "bench");
  g_signal_connect(window, "map", G_CALLBACK(mapped), app);
  gtk_application_add_window(app, GTK_WINDOW(window));
  gtk_window_present(GTK_WINDOW(window));
}

int main(void) {
  const char *name = getenv("BENCH_CASE");
  if (name == NULL)
    name = "";
  if (strcmp(name, "startup") == 0) {
    GtkApplication *app = gtk_application_new("dev.nts.Bench", G_APPLICATION_NON_UNIQUE);
    g_signal_connect(app, "activate", G_CALLBACK(activate), NULL);
    char *argv[] = {"bench", NULL};
    int status = g_application_run(G_APPLICATION(app), 1, argv);
    g_object_unref(app);
    return status;
  }
  gtk_init();
  if (strcmp(name, "signal") == 0) {
    adjustment = g_object_ref_sink(gtk_adjustment_new(0, 0, 1e12, 0, 0, 0));
    g_signal_connect(adjustment, "value-changed", G_CALLBACK(changed), NULL);
    best("signal", 500000, signal_run);
  } else if (strcmp(name, "property") == 0) {
    label = GTK_LABEL(g_object_ref_sink(gtk_label_new("a")));
    best("property", 500000, property_run);
  } else if (strcmp(name, "construct") == 0) {
    best("construct", 100000, construct_run);
  } else if (strcmp(name, "method") == 0) {
    button = g_object_ref_sink(gtk_button_new_with_label("x"));
    best("method", 2000000, method_run);
  } else if (strcmp(name, "vfunc") == 0) {
    square = g_object_ref_sink(g_object_new(bench_square_get_type(), NULL));
    best("vfunc", 200000, vfunc_run);
  } else if (strcmp(name, "outs") == 0) {
    button = g_object_ref_sink(gtk_button_new_with_label("x"));
    gtk_widget_set_size_request(button, 3, 4);
    best("outs", 2000000, outs_run);
  } else {
    printf("unknown case: %s\n", name);
  }
  return 0;
}

File: tooling/gtk-bench/gjs/bench.js
// The GJS half of the GTK benchmark; `../nts/src/main.ts` is the other, line
// for line. `BENCH_CASE` picks one case per process, and each prints
// `<case> <ns per operation>`: the best of three timed runs after one untimed,
// which the JIT needs.
imports.gi.versions.Gtk = '4.0';
const { GLib, GObject, Gio, Gtk } = imports.gi;

function now() {
    return GLib.get_monotonic_time() / 1e3;
}

function best(name, n, run) {
    run(n);
    let fastest = Infinity;
    for (let rep = 0; rep < 3; rep++) {
        const start = now();
        run(n);
        const took = now() - start;
        if (took < fastest)
            fastest = took;
    }
    print(`${name} ${(fastest * 1e6 / n).toFixed(1)}`);
}

function signal() {
    const adjustment = new Gtk.Adjustment({ upper: 1e12 });
    let changes = 0;
    let value = 0;
    adjustment.connect('value-changed', () => {
        changes++;
    });
    best('signal', 500000, n => {
        for (let i = 0; i < n; i++)
            adjustment.set_value(++value);
    });
    if (changes !== 500000 * 4)
        print(`signal: wrong count ${changes}`);
}

function property() {
    const label = new Gtk.Label({ label: 'a' });
    let length = 0;
    best('property', 500000, n => {
        for (let i = 0; i < n; i++) {
            label.label = (i & 1) === 0 ? 'a' : 'bb';
            length += label.label.length;
        }
    });
    if (length === 0)
        print('property: nothing read');
}

function construct() {
    best('construct', 100000, n => {
        for (let i = 0; i < n; i++) {
            const label = new Gtk.Label({ label: 'x' });
            if (label.label.length !== 1)
                print('construct: wrong label');
        }
    });
}

function method() {
    const button = new Gtk.Button({ label: 'x' });
    let visible = 0;
    best('method', 2000000, n => {
        for (let i = 0; i < n; i++)
            if (button.get_visible())
                visible++;
    });
    if (visible === 0)
        print('method: never visible');
}

function outs() {
    const button = new Gtk.Button({ label: 'x' });
    button.set_size_request(3, 4);
    let sum = 0;
    best('outs', 2000000, n => {
        for (let i = 0; i < n; i++) {
            const [width, height] = button.get_size_request();
            sum += width + height;
        }
    });
    if (sum === 0)
        print('outs: nothing read');
}

const Square = GObject.registerClass(class Square extends Gtk.Widget {
    vfunc_measure(orientation, forSize) {
        return [42, 42, -1, -1];
    }
});

function vfunc() {
    const square = new Square();
    let total = 0;
    best('vfunc', 200000, n => {
        for (let i = 0; i < n; i++) {
            const [minimum] = square.measure(Gtk.Orientation.HORIZONTAL, 100 + (i % 1000));
            total += minimum;
        }
    });
    if (total === 0)
        print('vfunc: never measured');
}

const Model = GObject.registerClass({Implements: [Gio.ListModel]}, class Model extends GObject.Object {
    vfunc_get_n_items() {
        return 7;
    }

    vfunc_get_item_type() {
        return GObject.Object.$gtype;
    }

    vfunc_get_item(position) {
        return null;
    }
});

function model() {
    const items = new Model();
    let total = 0;
    best('model', 2000000, n => {
        for (let i = 0; i < n; i++)
            total += items.get_n_items();
    });
    if (total === 0)
        print('model: never counted');
}

function startup() {
    const application = new Gtk.Application({ application_id: 'dev.nts.Bench', flags: Gio.ApplicationFlags.NON_UNIQUE });
    application.connect('activate', () => {
        const window = new Gtk.Window({ title: 'bench' });
        window.connect('map', () => {
            application.quit();
        });
        application.add_window(window);
        window.present();
    });
    application.run(['bench']);
}

const name = GLib.getenv('BENCH_CASE') ?? '';
if (name === 'startup') {
    startup();
} else {
    Gtk.init();
    if (name === 'signal')
        signal();
    else if (name === 'property')
        property();
    else if (name === 'construct')
        construct();
    else if (name === 'method')
        method();
    else if (name === 'outs')
        outs();
    else if (name === 'vfunc')
        vfunc();
    else if (name === 'model')
        model();
    else
        print(`unknown case: ${name}`);
}
